src/ir_Kelon.cpp

```diff
@@ -0,0 +1,499 @@
+/// @param[in] new_code The raw state from the native IR message.
+void IRKelonAc::setRaw(const uint64_t new_code) {
+  _.raw = new_code;
+}
+
+/// Convert a standard A/C mode (stdAc::opmode_t) into it a native mode.
+/// @param[in] mode A stdAc::opmode_t operation mode.
+/// @return The native mode equivalent.
+uint8_t IRKelonAc::convertMode(const stdAc::opmode_t mode) {
+  switch (mode) {
+    case stdAc::opmode_t::kCool:
+      return kKelonModeCool;
+    case stdAc::opmode_t::kHeat:
+      return kKelonModeHeat;
+    case stdAc::opmode_t::kDry:
+      return kKelonModeDry;
+    case stdAc::opmode_t::kFan:
+      return kKelonModeFan;
+    default:
+      return kKelonModeSmart;
+  }
+}
+
+/// Convert a standard A/C fan speed (stdAc::fanspeed_t) into it a native speed.
+/// @param[in] fan A stdAc::fanspeed_t fan speed
+/// @return The native speed equivalent.
+uint8_t IRKelonAc::convertFan(stdAc::fanspeed_t fan) {
+  switch (fan) {
+    case stdAc::fanspeed_t::kMin:
+    case stdAc::fanspeed_t::kLow:
+      return kKelonFanMin;
+    case stdAc::fanspeed_t::kMedium:
+      return kKelonFanMedium;
+    case stdAc::fanspeed_t::kHigh:
+    case stdAc::fanspeed_t::kMax:
+      return kKelonFanMax;
+    default:
+      return kKelonFanAuto;
+  }
+}
+
+/// Convert a native mode to it's stdAc::opmode_t equivalent.
+/// @param[in] mode A native operating mode value.
+/// @return The stdAc::opmode_t equivalent.
+stdAc::opmode_t IRKelonAc::toCommonMode(const uint8_t mode) {
+  switch (mode) {
+    case kKelonModeCool:
+      return stdAc::opmode_t::kCool;
+    case kKelonModeHeat:
+      return stdAc::opmode_t::kHeat;
+    case kKelonModeDry:
+      return stdAc::opmode_t::kDry;
+    case kKelonModeFan:
+      return stdAc::opmode_t::kFan;
+    default:
+      return stdAc::opmode_t::kAuto;
+  }
+}
+
+/// Convert a native fan speed to it's stdAc::fanspeed_t equivalent.
+/// @param[in] speed A native fan speed value.
+/// @return The stdAc::fanspeed_t equivalent.
+stdAc::fanspeed_t IRKelonAc::toCommonFanSpeed(const uint8_t speed) {
+  switch (speed) {
+    case kKelonFanMin:
+      return stdAc::fanspeed_t::kLow;
+    case kKelonFanMedium:
+      return stdAc::fanspeed_t::kMedium;
+    case kKelonFanMax:
+      return stdAc::fanspeed_t::kHigh;
+    default:
+      return stdAc::fanspeed_t::kAuto;
+  }
+}
+
+/// Convert the internal A/C object state to it's stdAc::state_t equivalent.
+/// @return A stdAc::state_t containing the current settings.
+stdAc::state_t IRKelonAc::toCommon(const stdAc::state_t *prev) const {
+  stdAc::state_t result{};
+  result.protocol = decode_type_t::KELON;
+  result.model = -1;  // Unused.
+  result.mode = toCommonMode(getMode());
+  result.celsius = true;
+  result.degrees = getTemp();
+  result.fanspeed = toCommonFanSpeed(getFan());
+  result.turbo = getSupercool();
+  result.sleep = getSleep() ? 0 : -1;
+  // Not supported.
+  // N/A, AC only supports toggling it
+  result.power = prev == nullptr || prev->power;
```

Comment:
Not sure, but isn't it:
```
result.power = ((prev == nullptr) || prev->power) ^ _.PowerToggle;
```
The power toggle should be included _somewhere_, I know that much! ;-)

Reply:
If prev->power is handled by the handleToggles routine, isn't it effectively the power toggle?

This reports it on if there's no previous value, since we have no way to know if it's actually on.

Reply:
> isn't it effectively the power toggle?
Answer is: Yes & No, err .. it's complicated. ;-)

`.toCommon()` isn't used by the `IRac` object so, no. `handleToggles()` is only used _internally_ by the `IRac` object. i.e. Between sending a state, and then sending the next desired state. `handleToggles()` doesn't change the state, it just calculates what state to send.

`.toCommon()` is mostly used by systems that _capture_ an IR message from a remote and wish to update an existing `IRac` object. It's a _state_ that is coming in externally to the object. See `IRMQTTServer` and Tasmota which use/do this. Here is the relevant code from `IRMQTTServer`:
https://github.com/crankyoldgit/IRremoteESP8266/blob/371b82e2058ef3a9a1929372cd7141ba50324472/examples/IRMQTTServer/IRMQTTServer.ino#L3139-L3186

The TL;DR: is this: `toCommon()` et al, is usual used to convert an incoming IR message into a IRac _state_. The _states_ don't know about toggles, thus the logic for toggles needs to be done elsewhere to _convert_ it into a sort-of common/universal state. This gets more complicated when you factor in, it is possible/desirable by some to control an A/C of one brand/protocol using a remote from another A/C brand/protocol. That said, we can't assume that is how someone is going to use it, hence the `prev` is optional.

Or to put it another way, using examples:
Assume your A/C is Off, having been set that way via you instructing an `IRac` object to make it so.
i.e. `IRac.next.power = false; IRac.send();` etc etc
Now, you press the "power" button on the remote, which sends a "power toggle" IR message.
You want the ESP to do the right thing, i.e. Toggle the power from "Off" to "On" (`next.power = true;`)
So, you need to convert the captured IR message (via `toCommon()`) to a _state_. All IRKelonAc can reliably determine is that it's a power toggle message, hence `toCommon()` needs to know the previous state to work out if the new _desired_ state is "On" or "Off". So, supplying it the previous power state (`prev->power`) xor'ed with `_.toggle_power` should give the desired _new_ power state.
i.e. "Yo library! I just got a power toggle message, what do I do?", "Well, it was in `power off`, so I guess it should be `power on` now.", "Cool, thanks!"

Have I made it clearer, or made it worse? ;-)

Reply:
Thanks for the additional context :)

`<unrequested-opinion>`

I feel, though, that all this work to handle states for things that can only be toggled is not super useful. If you start from a good state, cool.

However, since you're just guessing the initial state, you still have a 50% chance the reported state will still be wrong, which makes it pretty useless.

On top of that, even when starting from a state that matches the real-world state, at some point the AC or the ESP receiver will miss one signal and you're back to the initial problem.

I don't have a solution to propose for this and I don't think there's any that requires hardware-modding the AC unit or just getting a better one, but yeah I think trying to handle this case adds unnecessary complexity to the overall implementation.

In my specific case I learned from the user manual that going through a set of operations you can effectively assert the on/off status, and that's why I implemented `ensurePower()`. I'm keeping track of the power state in an external class in my code, which allows the user to adjust an incorrect initial state.

`<unrequested-opinion/>`

No problem, I'll add the xor condition

Reply:
> <unrequested-opinion>

Oh I almost 100% agree. No argument what so ever. I really dislike the protocols that _toggle_ and/or send special unique messages to change the state. It makes my life/this library's job a real pain. I can't think of a better way either, otherwise I'd have implemented it. It's the worst solution, except for all the other ones we've tried. ;-)
I don't even recommend the use-scenario. Having two controlling devices maintain the state of an A/C is asking for trouble. I wouldn't use it personally, but some people do. So I try to accommodate them as best as I can.

And yes, the next A/C system I buy, it will be one where the `state` transmitted contains _all_ the information. Toggles can die in a fire! ;-)